sway-lib-std/src/u128.sw

```diff
@@ -0,0 +1,310 @@
+library u128;
+
+use core::num::*;
+use ::assert::assert;
+use ::result::Result;
+
+/// The 128-bit unsigned integer type.
+/// Represented as two 64-bit components: `(upper, lower)`, where `value = (upper << 64) + lower`.
+pub struct U128 {
+    upper: u64,
+    lower: u64,
+}
+
+pub trait From {
+    /// Function for creating U128 from its u64 components.
+    pub fn from(h: u64, l: u64) -> Self;
+} {
+    fn into(v: U128) -> (u64, u64) {
+        (v.upper, v.lower)
+    }
+}
+
+impl From for U128 {
+    pub fn from(h: u64, l: u64) -> U128 {
+        U128 {
+            upper: h,
+            lower: l,
+        }
+    }
+}
+
+impl core::ops::Eq for U128 {
+    pub fn eq(self, other: Self) -> bool {
+        self.lower == other.lower && self.upper == other.upper
+    }
+}
+
+impl core::ops::Ord for U128 {
+    pub fn gt(self, other: Self) -> bool {
+        self.upper > other.upper || self.upper == other.upper && self.lower > other.lower
+    }
+
+    pub fn lt(self, other: Self) -> bool {
+        self.upper < other.upper || self.upper == other.upper && self.lower < other.lower
+    }
+}
+
+// TODO this doesn't work?
+// impl core::ops::OrdEq for U128 {
+// }
+
+fn disable_overflow() {
+    asm(r1) {
+        movi r1 i3;
+        flag r1;
+    }
+}
+
+fn enable_overflow() {
+    asm(r1) {
+        movi r1 i0;
+        flag r1;
+    }
+}
+
+impl u64 {
+    pub fn overflowing_add(a: u64, b: u64) -> U128 {
+        disable_overflow();
+        let mut v = U128 {
+            upper: 0,
+            lower: 0,
+        };
+        asm(r1, r2, a: a, b: b, v_ptr: v) {
+            add r1 a b;
+            move r2 of;
+            sw v_ptr r2 i0;
+            sw v_ptr r1 i1;
+        };
+        enable_overflow();
+        v
+    }
+
+    pub fn overflowing_mul(self, b: Self) -> U128 {
+        disable_overflow();
+        let mut v = U128 {
+            upper: 0,
+            lower: 0,
+        };
+        asm(r1, r2, a: self, b: b, v_ptr: v) {
+            mul r1 a b;
+            move r2 of;
+            sw v_ptr r2 i0;
+            sw v_ptr r1 i1;
+        };
+        enable_overflow();
+        v
+    }
+}
+
+impl U128 {
+    /// Initializes a new, zeroed U128.
+    pub fn new() -> U128 {
+        U128 {
+            upper: 0,
+            lower: 0,
+        }
+    }
+
+    /// Downcast to `u64`. Err if precision would be lost, Ok otherwise.
+    pub fn to_u64(self) -> Result<u64, ()> {
+        match self.upper {
+            0 => {
+                Result::Ok(self.lower)
+            },
+            _ => {
+                Result::Err(())
+            },
+        }
+    }
+
+    /// The smallest value that can be represented by this integer type.
+    pub fn min() -> U128 {
```

Comment:
Can we be cheeky and use the `new()` function from above?

Reply:
No it can't. #1548